# Assignments: Loops and Control Structures

**Name:** ___________________

**Date:** ___________________

---

## Instructions
- Complete all the problems below
- Write your code in the provided cells
- Test your solutions with different inputs
- Comment your code where necessary
- Each problem has a difficulty level: Easy, Medium, or Hard

---

## Section 1: Control Structures (If-Else Statements)

### Problem 1.1: Grade Calculator (Easy)

Write a program that takes a student's marks (0-100) as input and prints their grade according to the following criteria:
- 90-100: Grade A
- 80-89: Grade B
- 70-79: Grade C
- 60-69: Grade D
- Below 60: Grade F

Also, validate that the marks are within the valid range (0-100).

**Hint:** Use if-elif-else statements. First check if marks are in valid range (0-100), then use >= comparisons starting from highest grade.

In [125]:
# Your code here
marks = int(input("Enter the student's marks : "))

if marks >= 90 and marks <= 100:
    print(f"The student got {marks} marks  with Grade A") 
elif marks>=80 and marks <= 89:
    print(f"The student got {marks} marks  with Grade B")
elif marks>=70 and marks <= 79:
    print(f"The student got {marks} marks  with Grade C")
elif marks>=60 and marks <= 69:
    print(f"The student got {marks} marks  with Grade D")
else :
    print(f"The student got {marks} marks  with Grade F")

The student got 12345 marks  with Grade F


### Problem 1.2: Leap Year Checker (Medium)

Write a program to check if a given year is a leap year or not.

Rules:
- A year is a leap year if it is divisible by 4
- However, if it is divisible by 100, it is NOT a leap year
- Exception: If it is also divisible by 400, then it IS a leap year

Examples:
- 2000: Leap year (divisible by 400)
- 1900: Not a leap year (divisible by 100 but not 400)
- 2024: Leap year (divisible by 4 but not 100)

**Hint:** Check conditions in order: first divisible by 400, then by 100, then by 4. Use modulo operator (%) to check divisibility.

In [ ]:
# Your code here
year = int(input("Enter the year: "))

if (year % 4 == 0 and year % 100 != 0) or year % 400 ==0 :
    print(f"{year} is a leap year")

else:
     print(f"{year} is not a leap year")

### Problem 1.3: Triangle Type Identifier (Medium)

Write a program that takes three sides of a triangle as input and determines:
1. Whether the three sides can form a valid triangle (sum of any two sides must be greater than the third side)
2. If valid, classify the triangle as:
   - Equilateral (all three sides equal)
   - Isosceles (exactly two sides equal)
   - Scalene (no sides equal)

**Hint:** First check: (a+b>c) and (b+c>a) and (a+c>b). Then use == to compare sides: if all three equal, it's equilateral; if any two equal, it's isosceles; otherwise scalene.

In [ ]:
# Your code here
a = float(input("Enter side a: "))
b = float(input("Enter side b: "))
c = float(input("Enter side c: "))

if (a+b>c) == (b+c>a) == (a+c>b):
     print("The sides form a valid triangle.")

     if a == b == c:
        print("It is an Equilateral triangle.")
     elif a == b or b == c or a == c :
        print("It is an Isosceles triangle.")
     else:
        print("It is a Scalene triangle.")
else:
     print("The sides do not form a valid triangle.")

The sides form a valid triangle.
It is a Scalene triangle.


### Problem 1.4: BMI Calculator with Categories (Easy)

Write a program that calculates BMI (Body Mass Index) and categorizes it.

Formula: BMI = weight (kg) / (height (m))²

Categories:
- Below 18.5: Underweight
- 18.5-24.9: Normal weight
- 25-29.9: Overweight
- 30 and above: Obese

**Hint:** First calculate BMI using the formula (use ** for power or multiply height by itself). Then use if-elif-else to categorize.

In [ ]:
# Your code here
weight = float(input("Enter the weight in kgs: "))
height = float(input("Enter the height in mts:"))
bmi = weight / (height)**2

if bmi >= 30:
    print("The body has obese")

elif bmi>=25 and bmi<=29.9:
     print("The body has overweight")

elif bmi>=18 and bmi<=24.9:
     print("The body has Normal Weight")

else:
     print("The body has under Weight")


The body has under Weight


---

## Section 2: For Loops

### Problem 2.1: Multiplication Table (Easy)

Write a program that prints the multiplication table for a given number (1 to 10).

Example output for number 5:
```
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
...
5 x 10 = 50
```

**Hint:** Use for loop with range(1, 11) to iterate from 1 to 10. Multiply the number by loop variable i.

In [ ]:
# Your code here
number = int(input("Enter a number : "))

for i in range(1,11):
    print(f"{number} x {i} = {number * i}")

6 x 1 = 6
6 x 2 = 12
6 x 3 = 18
6 x 4 = 24
6 x 5 = 30
6 x 6 = 36
6 x 7 = 42
6 x 8 = 48
6 x 9 = 54
6 x 10 = 60


### Problem 2.2: Factorial Calculator (Medium)

Write a program to calculate the factorial of a number using a for loop.

Factorial of n (n!) = n × (n-1) × (n-2) × ... × 1

Example: 5! = 5 × 4 × 3 × 2 × 1 = 120

**Hint:** Initialize result = 1. Use for loop with range(1, n+1) and multiply result by each number.

In [ ]:
# Your code here
n = int(input("Enter a factorial number: "))
result = 1
for i in range(1,n+1):
    result *= i

print(f"The factiorial number of {n} is {result}")

The factiorial number of 5 is 120


### Problem 2.3: Sum of Even Numbers (Easy)

Write a program that calculates the sum of all even numbers between 1 and n (inclusive), where n is provided by the user.

Example: If n = 10, sum = 2 + 4 + 6 + 8 + 10 = 30

**Hint:** Use for loop with range(1, n+1). Check if i % 2 == 0, then add to sum. Or use range(2, n+1, 2) to iterate only even numbers.

In [ ]:
# Your code here
n = int(input("Enter a  number: "))
sum = 0
for i in range(1,n+1):
    if i%2 == 0:
        sum += i
print(f"Sum of even numbers upto {n} is {sum}")

Sum of even numbers upto 10 is 30


### Problem 2.4: Fibonacci Series (Medium)

Write a program to print the first n terms of the Fibonacci series.

The Fibonacci series: 0, 1, 1, 2, 3, 5, 8, 13, 21, ...

Each number is the sum of the two preceding ones.

**Hint:** Initialize a=0, b=1. Print a, then in loop: calculate next=a+b, print next, then update a=b and b=next.

In [ ]:
# Your code here
n = int(input("Enter a  number: "))
a = 0
b = 1
print("fibonacci Series")
for i in range(n):
    print(a, end=" ")
    # a, b = b, a+b
    next = a+b
    a = b
    b = next

fibonacci Series
0 1 1 2 3 5 8 13 21 34 

### Problem 2.5: Prime Number Checker (Medium)

Write a program to check if a given number is prime or not.

A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself.

**Hint:** Check if n is divisible by any number from 2 to n-1. If no divisor found, it's prime. Optimize: check only up to √n.

In [ ]:
# Your code here
n = int(input("Enter a number: "))

if n > 1:
    is_prime = True
    for i in range(2, n):  
        if n % i == 0:
            is_prime = False
            break

    if is_prime:
        print(n, "is a prime number")
    else:
        print(n, "is not a prime number")
else:
    print(n, "is not a prime number")


3 is a prime number


### Problem 2.6: Count Digits in a Number (Easy)

Write a program that counts the number of digits in a given integer.

Example: 12345 has 5 digits

**Hint:** Convert number to string and use len(), or keep dividing by 10 until number becomes 0, counting each division.

In [ ]:
# Your code here
number = int(input("Enter a number: "))
count = 0

for i in str(abs(number)):
    count += 1

print(f"{number} has {count} digits")   

12365 has 5 digits


---

## Section 3: While Loops

### Problem 3.1: Reverse a Number (Medium)

Write a program that reverses the digits of a given number using a while loop.

Example: Input: 12345, Output: 54321

**Hint:** Initialize reverse=0. While n>0: get last digit (n%10), add to reverse (reverse=reverse*10+digit), remove last digit (n=n//10).

In [5]:
# Your code here
num1 = int(input("Enter a number : "))
rev = 0
temp = num1
while num1>0:
    rev = rev*10 + num1%10
    num1 //= 10
print(f"Before number : {temp}")
print(f"After Reversed :{rev}")

Before number : 123456
After Reversed :654321


### Problem 3.2: Sum of Digits (Easy)

Write a program that calculates the sum of all digits in a given number.

Example: Input: 12345, Output: 15 (1+2+3+4+5)

**Hint:** While n>0: add last digit to sum (sum += n%10), then remove it (n = n//10).

In [ ]:
# Your code here
num2 = int(input("Enter a number: "))
original = num2
sum = 0
n = num2

while n>0:
    sum += n%10
    n//=10
print(f"Sum of digits this number {original} is {sum} ")

Sum of digits this number123456 is 21 


### Problem 3.3: Number Guessing Game (Medium)

Write a number guessing game where:
1. The program generates a random number between 1 and 100
2. The user has to guess the number
3. After each guess, the program tells the user if their guess is too high or too low
4. The game continues until the user guesses correctly
5. Display the number of attempts taken

**Hint:** Use `import random` and `random.randint(1, 100)` to generate number. Use while loop with condition guess != target. Count attempts.

In [16]:
# Your code here
import random

target  = random.randint(1,100)

attempt = 0
guess = None

while guess != target:
    
    guess = int(input("Enter your guess: "))
    print("Enter your guess : ",guess)
    attempt += 1

    if guess < target:
        print("Too low! Try again.")
    elif guess > target:
        print("Too high! Try again.")
    else:
        print(f"Correct! The number was {target}.")
        print(f"You guessed it in {attempt} attempts.")

Too high! Try again.
Too high! Try again.
Too high! Try again.
Too high! Try again.
Too low! Try again.
Too low! Try again.
Too low! Try again.
Too high! Try again.
Too high! Try again.
Too high! Try again.
Too high! Try again.
Correct! The number was 16.
You guessed it in 12 attempts.


### Problem 3.4: Palindrome Checker (Medium)

Write a program to check if a given number is a palindrome or not.

A palindrome number reads the same forwards and backwards.

Examples: 121, 12321, 1331 are palindromes

**Hint:** Store original number, reverse it (like Problem 3.1), then compare original with reversed number.

In [19]:
# Your code here
num1 = int(input("Enter a number : "))
rev = 0
temp = num1
while num1>0:
    rev = rev*10 + num1%10
    num1 //= 10

if rev == temp:
        print(f"{temp} is a palindrome. ")
else:
        print(f"{temp} is not a palindrome. ")



121 is a palindrome. 


### Problem 3.5: Armstrong Number (Hard)

Write a program to check if a number is an Armstrong number (also called Narcissistic number).

An Armstrong number is a number that is equal to the sum of its own digits each raised to the power of the number of digits.

Examples:
- 153 = 1³ + 5³ + 3³ = 1 + 125 + 27 = 153 (Armstrong number)
- 9474 = 9⁴ + 4⁴ + 7⁴ + 4⁴ = 6561 + 256 + 2401 + 256 = 9474 (Armstrong number)

**Hint:** First count digits (power). Store original. Extract each digit, raise to power, add to sum. Compare sum with original.

In [26]:
# Your code here
num_1 = int(input("Enter a number: "))
original = num_1
armstrong_number  = 0

while num_1 > 0 :
    digit = num_1 % 10
    armstrong_number += digit**3 
    num_1 //= 10

if original == armstrong_number:
    print(f"{original} is an armstrong number ")
else:
    print(f"{original} is not an armstrong number ")

153 is an armstrong number 


---

## Section 4: Nested Loops

### Problem 4.1: Pattern Printing - Right Triangle (Easy)

Print the following pattern for n rows:

```
*
**
***
****
*****
```
(Example shown for n=5)

**Hint:** Use nested loops: outer loop for rows (i from 1 to n), inner loop for columns (j from 1 to i). Print * in inner loop.

In [27]:
# Your code
n = int(input("Enter a number"))
for i in range(n+1):
    for j in range(i):
        print("*", end=" ")
    print()


* 
* * 
* * * 
* * * * 
* * * * * 


### Problem 4.2: Pattern Printing - Inverted Right Triangle (Easy)

Print the following pattern for n rows:

```
*****
****
***
**
*
```
(Example shown for n=5)

**Hint:** Outer loop for rows. Inner loop prints stars from 1 to (n-i+1), where i is the row number.

In [34]:
# Your code here
n = int(input("Enter a number"))
for i in range(n,0,-1):
    for j in range(i):
        print("*", end=" ")
    print()

* * * * * 
* * * * 
* * * 
* * 
* 


### Problem 4.3: Pattern Printing - Pyramid (Medium)

Print the following pyramid pattern for n rows:

```
    *
   ***
  *****
 *******
*********
```
(Example shown for n=5)

**Hint:** For row i: print (n-i) spaces, then print (2*i-1) stars. Use nested loops or string multiplication.

In [38]:
# Your code here

n = int(input("Enter a number : "))
for i in range(1,n+1):
    for j in range(n-i):
        print(" ",end=" ")
    for k in range(2 * i - 1):
     print("*", end=" ")
    
    print()

        * 
      * * * 
    * * * * * 
  * * * * * * * 
* * * * * * * * * 


### Problem 4.4: Pattern Printing - Number Pattern (Medium)

Print the following number pattern for n rows:

```
1
12
123
1234
12345
```
(Example shown for n=5)

**Hint:** Similar to Problem 4.1, but print j (column number) instead of *.

In [43]:
# Your code here
n = int(input("Enter a number : "))

for i in range(1,n+1):
    for j in range(1,i+1):
        print(j,end=" ")
    print()

1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


### Problem 4.5: Multiplication Table Grid (Medium)

Print a multiplication table grid from 1 to n.

Example for n=5:
```
1   2   3   4   5
2   4   6   8  10
3   6   9  12  15
4   8  12  16  20
5  10  15  20  25
```

**Hint:** Nested loops: for i in range(1, n+1) and for j in range(1, n+1). Print i*j with proper spacing.

In [6]:
# Your code here
table = int(input("Enter a number for multiplication table: "))
for i in range(1, table + 1):
    for j in range(1, table + 1):
         print(f"{i * j:>3}", end=" ")
    print()


  1   2   3   4   5 
  2   4   6   8  10 
  3   6   9  12  15 
  4   8  12  16  20 
  5  10  15  20  25 


### Problem 4.6: Diamond Pattern (Hard)

Print a diamond pattern for n rows (n should be odd):

```
    *
   ***
  *****
 *******
*********
 *******
  *****
   ***
    *
```
(Example shown for n=9)

**Hint:** Print pyramid (upper half) for rows 1 to (n+1)//2, then print inverted pyramid (lower half) for remaining rows.

In [8]:
# Your code here
n = int(input("Enter a number : "))
for i in range(1,(n+1) // 2 + 1):
    for j in range((n//2)-(i-1)):
        print(" ",end=" ")
    for k in range(2 * i - 1):
     print("*", end=" ")
    
    print()

for i in range((n+1) // 2-1, 0, -1):
    for j in range((n//2)-(i-1)):
        print(" ",end=" ")
    for k in range(2 * i - 1):
     print("*", end=" ")
    
    print()


    * 
  * * * 
* * * * * 
  * * * 
    * 


---

## Section 5: Break and Continue Statements

### Problem 5.1: Find First Multiple (Easy)

Write a program that finds and prints the first number in the range 1 to 100 that is divisible by both 7 and 13. Use the `break` statement to exit the loop once found.

**Hint:** Loop through 1 to 100. Check if i%7==0 and i%13==0. If true, print i and use break to exit loop.

In [1]:
# Your code here
for i in range(1, 101):
    if i % 7 == 0 and i % 13 == 0:
        print(i)
        break

91


### Problem 5.2: Skip Multiples (Easy)

Write a program that prints all numbers from 1 to 50, but skips (doesn't print) numbers that are multiples of 3. Use the `continue` statement.

**Hint:** Loop from 1 to 50. If i%3==0, use continue to skip to next iteration. Otherwise print i.

In [2]:
# Your code here
for i in range(1, 51):
    if i % 3 == 0:
        continue
    print(i)

1
2
4
5
7
8
10
11
13
14
16
17
19
20
22
23
25
26
28
29
31
32
34
35
37
38
40
41
43
44
46
47
49
50


### Problem 5.3: Sum Until Negative (Medium)

Write a program that:
1. Keeps asking the user to enter numbers
2. Adds positive numbers to a sum
3. Skips zero (continues to next iteration)
4. Stops when a negative number is entered
5. Prints the final sum

**Hint:** Use while True loop. If num < 0, break. If num == 0, continue. Otherwise add to sum.

In [3]:
# Your code here
total = 0

while True:
    num = int(input("Enter a number: "))

    if num < 0:
        break

    if num == 0:
        continue

    total += num

print("Final sum:", total)


Final sum: 62


---

## Section 6: Combined Challenges

### Problem 6.1: Prime Numbers in Range (Medium)

Write a program to print all prime numbers between two given numbers (inclusive).

Example: Between 10 and 30: 11, 13, 17, 19, 23, 29

**Hint:** Outer loop for each number in range. For each number, use inner loop to check if it's prime (like Problem 2.5).

In [4]:
# Your code here
start = int(input("Enter starting number: "))
end = int(input("Enter ending number: "))

for num in range(start, end + 1):

    if num < 2:
        continue

    is_prime = True

    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            is_prime = False
            break

    if is_prime:
        print(num, end=" ")

11 13 17 19 23 29 

### Problem 6.2: GCD Calculator (Medium)

Write a program to find the Greatest Common Divisor (GCD) of two numbers using the Euclidean algorithm.

Algorithm:
1. Divide the larger number by the smaller number
2. Replace the larger number with the smaller number
3. Replace the smaller number with the remainder
4. Repeat until the remainder is 0
5. The GCD is the last non-zero remainder

**Hint:** While b != 0: store remainder (a%b), then a=b, b=remainder. When loop ends, a is the GCD.

In [5]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))

while b != 0:
    remainder = a % b
    a = b
    b = remainder

print("GCD:", a)

GCD: 10


### Problem 6.3: Perfect Number Checker (Hard)

Write a program to check if a number is a perfect number.

A perfect number is a positive integer that is equal to the sum of its proper positive divisors (excluding the number itself).

Example: 6 is perfect because 6 = 1 + 2 + 3

Example: 28 is perfect because 28 = 1 + 2 + 4 + 7 + 14

**Hint:** Loop from 1 to n-1. If i divides n (n%i==0), add i to sum. Compare sum with n.

In [6]:
# Your code here
n = int(input("Enter a number: "))

total = 0

for i in range(1, n):
    if n % i == 0:
        total += i

if total == n:
    print("Perfect number")
else:
    print("Not a perfect number")

Not a perfect number


### Problem 6.4: LCM Calculator (Medium)

Write a program to find the Least Common Multiple (LCM) of two numbers.

Hint: LCM(a, b) = (a × b) / GCD(a, b)

You can use the GCD function from Problem 6.2.

**Hint:** First calculate GCD using Euclidean algorithm (Problem 6.2), then use formula: LCM = (a * b) // GCD.

In [7]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))

x = a
y = b

while y != 0:
    remainder = x % y
    x = y
    y = remainder

gcd = x

lcm = (a * b) // gcd

print("LCM:", lcm)

LCM: 60


### Problem 6.5: Number to Words (Hard)

Write a program that converts a number (0-999) to words.

Examples:
- 0: Zero
- 25: Twenty Five
- 123: One Hundred Twenty Three
- 500: Five Hundred

**Hint:** Create lists for ones, teens, tens, etc. Extract hundreds, tens, ones digits. Convert each part to words and combine.

In [8]:
# Your code here
ones = [
    "Zero", "One", "Two", "Three", "Four",
    "Five", "Six", "Seven", "Eight", "Nine"
]

teens = [
    "Ten", "Eleven", "Twelve", "Thirteen", "Fourteen",
    "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"
]

tens = [
    "", "", "Twenty", "Thirty", "Forty",
    "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
]

n = int(input("Enter a number (0-999): "))

if n < 0 or n > 999:
    print("Invalid number")

elif n == 0:
    print("Zero")

else:
    result = ""

    if n >= 100:
        result += ones[n // 100] + " Hundred"
        n = n % 100

        if n > 0:
            result += " "

    if n >= 20:
        result += tens[n // 10]
        n = n % 10

        if n > 0:
            result += " " + ones[n]

    elif n >= 10:
        result += teens[n - 10]

    elif n > 0:
        result += ones[n]

    print(result)


Three Hundred Thirty Three


### Problem 6.6: Strong Number Checker (Hard)

Write a program to check if a number is a Strong number.

A Strong number is a number whose sum of factorial of digits is equal to the original number.

Example: 145 is a strong number because 1! + 4! + 5! = 1 + 24 + 120 = 145

**Hint:** Extract each digit, calculate its factorial, add to sum. Compare sum with original number.

In [9]:
# Your code here
n = int(input("Enter a number: "))

original = n
total = 0

while n > 0:
    digit = n % 10

    factorial = 1

    for i in range(1, digit + 1):
        factorial *= i

    total += factorial
    n //= 10

if total == original:
    print("Strong number")
else:
    print("Not a strong number")


Not a strong number


### Problem 6.7: Pascal's Triangle (Hard)

Write a program to print Pascal's Triangle up to n rows.

Example for n=5:
```
    1
   1 1
  1 2 1
 1 3 3 1
1 4 6 4 1
```

Each number is the sum of the two numbers directly above it.

**Hint:** Use formula: C(n,k) = C(n,k-1) * (n-k+1) / k, where C(n,0)=1. Or use nested loops with list to store previous row.

In [10]:
# Your code here
n = int(input("Enter number of rows: "))

for i in range(n):
    value = 1

    print(" " * (n - i), end="")

    for j in range(i + 1):
        print(value, end=" ")

        value = value * (i - j) // (j + 1)

    print()

  1 
 1 1 


### Problem 6.8: Simple Calculator (Medium)

Create a simple calculator program that:
1. Shows a menu of operations: Add, Subtract, Multiply, Divide, Exit
2. Takes two numbers as input
3. Performs the selected operation
4. Shows the result
5. Repeats until the user chooses to exit
6. Handles division by zero error

**Hint:** Use while loop. Display menu, get choice. Use if-elif to perform operation. Check if divisor is 0 before division.

In [11]:
# Your code here
while True:

    print("\n--- Calculator ---")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")
    print("5. Exit")

    choice = int(input("Enter your choice: "))

    if choice == 5:
        print("Calculator closed")
        break

    if choice >= 1 and choice <= 4:

        num1 = float(input("Enter first number: "))
        num2 = float(input("Enter second number: "))

        if choice == 1:
            print("Result:", num1 + num2)

        elif choice == 2:
            print("Result:", num1 - num2)

        elif choice == 3:
            print("Result:", num1 * num2)

        elif choice == 4:
            if num2 == 0:
                print("Cannot divide by zero")
            else:
                print("Result:", num1 / num2)

    else:
        print("Invalid choice")


--- Calculator ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Invalid choice

--- Calculator ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Result: 23.0

--- Calculator ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Result: -220.0

--- Calculator ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Calculator closed


---

## Section 7: Bonus Challenges

### Problem 7.1: Collatz Conjecture (Medium)

The Collatz conjecture is one of the most famous unsolved problems in mathematics. The rules are:
1. Start with any positive integer n
2. If n is even, divide it by 2
3. If n is odd, multiply it by 3 and add 1
4. Repeat until you reach 1

Write a program that:
- Takes a number as input
- Shows the complete sequence until it reaches 1
- Counts how many steps it took

Example: Starting with 6: 6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1 (8 steps)

**Hint:** While n != 1: if n%2==0, n=n//2, else n=3*n+1. Count steps and print each value.

In [2]:
# Your code here
n = int(input("Enter a positive integer: "))

steps = 0

print("Sequence:")

while n != 1:
    print(n, end=" → ")

    if n % 2 == 0:
        n = n // 2
    else:
        n = 3 * n + 1

    steps += 1

print(1)
print("Number of steps:", steps)

Sequence:
6 → 3 → 10 → 5 → 16 → 8 → 4 → 2 → 1
Number of steps: 8


### Problem 7.2: Caesar Cipher (Hard)

Write a program that implements a Caesar cipher to encrypt and decrypt messages.

The Caesar cipher shifts each letter by a fixed number of positions in the alphabet.

Example: With shift=3
- 'A' becomes 'D'
- 'Hello' becomes 'Khoor'

Your program should:
1. Ask for a message and shift value
2. Provide options to encrypt or decrypt
3. Handle both uppercase and lowercase letters
4. Keep non-alphabetic characters unchanged

**Hint:** Loop through each character. Use ord() to get ASCII value, shift it, use chr() to convert back. Use modulo 26 for wrapping. Check if char.isalpha().

In [3]:
# Your code here
message = input("Enter message: ")
shift = int(input("Enter shift value: "))

choice = input("Enter E for Encrypt or D for Decrypt: ").upper()

result = ""

for ch in message:

    if ch.isalpha():

        if ch.isupper():
            start = ord('A')
        else:
            start = ord('a')

        if choice == "E":
            new_char = chr((ord(ch) - start + shift) % 26 + start)

        elif choice == "D":
            new_char = chr((ord(ch) - start - shift) % 26 + start)

        else:
            print("Invalid choice")
            break

        result += new_char

    else:
        result += ch

print("Result:", result)

Result: Olssv


### Problem 7.3: ATM Machine Simulator (Hard)

Create an ATM machine simulator with the following features:
1. Initial balance: 10000
2. PIN: 1234 (3 attempts allowed)
3. Menu options:
   - Check Balance
   - Withdraw Money (check for sufficient balance and valid amount)
   - Deposit Money
   - Change PIN
   - Exit
4. Keep transaction history (last 5 transactions)
5. Input validation for all operations

**Hint:** Use variables for balance, pin. Loop for PIN attempts. Use while loop for menu. Use list to store transaction history. Validate amounts > 0 and balance >= withdrawal.

In [4]:
# Your code here
balance = 10000
pin = 1234
history = []

attempts = 3

while attempts > 0:

    entered_pin = int(input("Enter your PIN: "))

    if entered_pin == pin:

        while True:

            print("\n--- ATM MENU ---")
            print("1. Check Balance")
            print("2. Withdraw Money")
            print("3. Deposit Money")
            print("4. Change PIN")
            print("5. Exit")

            choice = input("Enter your choice: ")

            if choice == "1":

                print("Balance:", balance)
                history.append(f"Balance checked: {balance}")

            elif choice == "2":

                amount = float(input("Enter withdrawal amount: "))

                if amount <= 0:
                    print("Invalid amount")

                elif amount > balance:
                    print("Insufficient balance")

                else:
                    balance -= amount
                    print("Withdrawal successful")
                    print("Remaining balance:", balance)

                    history.append(f"Withdrawn: {amount}")

            elif choice == "3":

                amount = float(input("Enter deposit amount: "))

                if amount <= 0:
                    print("Invalid amount")

                else:
                    balance += amount
                    print("Deposit successful")
                    print("Current balance:", balance)

                    history.append(f"Deposited: {amount}")

            elif choice == "4":

                new_pin = int(input("Enter new PIN: "))
                pin = new_pin

                print("PIN changed successfully")
                history.append("PIN changed")

            elif choice == "5":

                print("\nTransaction History:")

                for transaction in history[-5:]:
                    print(transaction)

                print("Thank you for using ATM")
                break

            else:
                print("Invalid choice")

        break

    else:

        attempts -= 1
        print("Incorrect PIN")
        print("Attempts remaining:", attempts)

        if attempts == 0:
            print("Card blocked")


--- ATM MENU ---
1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit
Withdrawal successful
Remaining balance: 7000.0

--- ATM MENU ---
1. Check Balance
2. Withdraw Money
3. Deposit Money
4. Change PIN
5. Exit

Transaction History:
Withdrawn: 3000.0
Thank you for using ATM


---

## Section 8: Additional If-Else Practice (30 Questions)

### Problem 8.1: Positive, Negative, or Zero (Easy)

Write a program that checks if a number is positive, negative, or zero.

**Hint:** Use if-elif-else: if n > 0 (positive), elif n < 0 (negative), else (zero).

In [5]:
# Your code here
n = int(input("Enter a number: "))

if n > 0:
    print("Positive")
elif n < 0:
    print("Negative")
else:
    print("Zero")

Positive


### Problem 8.2: Even or Odd (Easy)

Write a program that determines if a number is even or odd.

**Hint:** If n % 2 == 0, it's even; otherwise odd.

In [6]:
# Your code here
n = int(input("Enter a number: "))

if n % 2 == 0:
    print("Even")
else:
    print("Odd")

Even


### Problem 8.3: Largest of Three Numbers (Easy)

Write a program to find the largest of three numbers.

**Hint:** Compare: if a >= b and a >= c, largest is a. Use nested if-else or use max() function.

In [7]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
c = int(input("Enter third number: "))

if a >= b and a >= c:
    largest = a
elif b >= a and b >= c:
    largest = b
else:
    largest = c

print("Largest:", largest)

Largest: 333


### Problem 8.4: Voting Eligibility (Easy)

Check if a person is eligible to vote (age >= 18).

**Hint:** If age >= 18, print eligible; else not eligible.

In [8]:
# Your code here
age = int(input("Enter your age: "))

if age >= 18:
    print("Eligible to vote")
else:
    print("Not eligible to vote")

Eligible to vote


### Problem 8.5: Character Type Checker (Medium)

Write a program that checks if a character is a vowel, consonant, digit, or special character.

**Hint:** Check if ch in 'aeiouAEIOU' (vowel), elif ch.isalpha() (consonant), elif ch.isdigit() (digit), else special.

In [9]:
# Your code here
ch = input("Enter a character: ")

if ch in "aeiouAEIOU":
    print("Vowel")
elif ch.isalpha():
    print("Consonant")
elif ch.isdigit():
    print("Digit")
else:
    print("Special character")

Vowel


### Problem 8.6: Discount Calculator (Medium)

Calculate discount based on purchase amount:
- Below 1000: No discount
- 1000-5000: 10% discount
- 5000-10000: 20% discount
- Above 10000: 30% discount

**Hint:** Use if-elif-else with ranges. Calculate: final_amount = amount - (amount * discount_percentage / 100).

In [10]:
# Your code here
amount = float(input("Enter purchase amount: "))

if amount < 1000:
    discount = 0
elif amount < 5000:
    discount = 10
elif amount < 10000:
    discount = 20
else:
    discount = 30

discount_amount = amount * discount / 100
final_amount = amount - discount_amount

print("Discount:", discount, "%")
print("Final amount:", final_amount)

Discount: 20 %
Final amount: 4440.0


### Problem 8.7: Quadrant Finder (Medium)

Given x and y coordinates, determine which quadrant the point lies in (or on an axis).

**Hint:** If x>0 and y>0: Quadrant I. If x<0 and y>0: Quadrant II. Similar for III and IV. Check for axes when x==0 or y==0.

In [11]:
# Your code here
x = float(input("Enter x: "))
y = float(input("Enter y: "))

if x == 0 and y == 0:
    print("Origin")
elif x == 0:
    print("Y-axis")
elif y == 0:
    print("X-axis")
elif x > 0 and y > 0:
    print("Quadrant I")
elif x < 0 and y > 0:
    print("Quadrant II")
elif x < 0 and y < 0:
    print("Quadrant III")
else:
    print("Quadrant IV")

Quadrant I


### Problem 8.8: Age Category (Easy)

Categorize age as: Child (0-12), Teenager (13-19), Adult (20-59), Senior (60+).

**Hint:** Use if-elif-else with age ranges. Check from Child to Senior in order.

In [12]:
# Your code here
age = int(input("Enter age: "))

if 0 <= age <= 12:
    print("Child")
elif age <= 19:
    print("Teenager")
elif age <= 59:
    print("Adult")
else:
    print("Senior")

Adult


### Problem 8.9: Temperature Converter (Medium)

Convert temperature between Celsius and Fahrenheit based on user choice.
- C to F: F = (C × 9/5) + 32
- F to C: C = (F - 32) × 5/9

**Hint:** Ask user for choice (1 or 2). Use if-else to apply appropriate formula. Use parentheses correctly in calculations.

In [13]:
# Your code here
print("1. Celsius to Fahrenheit")
print("2. Fahrenheit to Celsius")

choice = int(input("Enter choice: "))
temp = float(input("Enter temperature: "))

if choice == 1:
    fahrenheit = (temp * 9 / 5) + 32
    print("Fahrenheit:", fahrenheit)

elif choice == 2:
    celsius = (temp - 32) * 5 / 9
    print("Celsius:", celsius)

else:
    print("Invalid choice")

1. Celsius to Fahrenheit
2. Fahrenheit to Celsius
Fahrenheit: 431.6


### Problem 8.10: Password Strength Checker (Medium)

Check password strength:
- Weak: length < 6
- Medium: length 6-10
- Strong: length > 10

**Hint:** Use len(password) to get length. Compare with if-elif-else.

In [14]:
password = input("Enter password: ")

length = len(password)

if length < 6:
    print("Weak")
elif length <= 10:
    print("Medium")
else:
    print("Strong")

Medium


### Problem 8.11: Divisibility Checker (Easy)

Check if a number is divisible by both 3 and 5, only 3, only 5, or neither.

**Hint:** Check if n%3==0 and n%5==0 (both), elif n%3==0 (only 3), elif n%5==0 (only 5), else (neither).

In [15]:
# Your code here
n = int(input("Enter a number: "))

if n % 3 == 0 and n % 5 == 0:
    print("Divisible by both 3 and 5")
elif n % 3 == 0:
    print("Divisible by 3 only")
elif n % 5 == 0:
    print("Divisible by 5 only")
else:
    print("Neither divisible by 3 nor 5")

Divisible by 3 only


### Problem 8.12: Movie Ticket Pricing (Medium)

Calculate ticket price:
- Children (< 12): $10
- Adults (12-59): $15
- Seniors (60+): $12
- Weekend: Add $3 to all prices

**Hint:** First determine base price by age. Then check if weekend and add $3 if true.

In [17]:
# Your code here
age = int(input("Enter age: "))
weekend = input("Is it weekend? (yes/no): ").lower()

if age < 12:
    price = 10
elif age < 60:
    price = 15
else:
    price = 12

if weekend == "yes":
    price += 3

print("Ticket price: $", price)

Ticket price: $ 15


### Problem 8.13: Letter Grade to GPA (Medium)

Convert letter grade to GPA: A=4.0, B=3.0, C=2.0, D=1.0, F=0.0

**Hint:** Use if-elif-else or dictionary: grade_dict = {'A': 4.0, 'B': 3.0, ...}. Then print grade_dict[letter].

In [18]:
# Your code here
grade = input("Enter grade: ").upper()

if grade == "A":
    gpa = 4.0
elif grade == "B":
    gpa = 3.0
elif grade == "C":
    gpa = 2.0
elif grade == "D":
    gpa = 1.0
elif grade == "F":
    gpa = 0.0
else:
    gpa = None
    print("Invalid grade")

if gpa is not None:
    print("GPA:", gpa)

GPA: 4.0


### Problem 8.14: Electricity Bill Calculator (Medium)

Calculate bill based on units consumed:
- 0-100: $0.50 per unit
- 101-200: $0.75 per unit
- 201-300: $1.20 per unit
- Above 300: $1.50 per unit

**Hint:** Use if-elif-else to determine rate based on units. Calculate: bill = units * rate.

In [19]:
# Your code here
units = float(input("Enter units consumed: "))

if units <= 100:
    rate = 0.50
elif units <= 200:
    rate = 0.75
elif units <= 300:
    rate = 1.20
else:
    rate = 1.50

bill = units * rate

print("Electricity bill:", bill)

Electricity bill: 832.5


### Problem 8.15: Traffic Light System (Easy)

Based on light color, print action: Red=Stop, Yellow=Caution, Green=Go

**Hint:** Use if-elif-else or dictionary to map color to action.

In [20]:
# Your code here
light = input("Enter light color: ").lower()

if light == "red":
    print("Stop")
elif light == "yellow":
    print("Caution")
elif light == "green":
    print("Go")
else:
    print("Invalid color")

Caution


### Problem 8.16: Smallest of Three Numbers (Easy)

Find the smallest of three numbers.

**Hint:** Similar to Problem 8.3, but check for smallest. Or use min(a, b, c).

In [21]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))
c = int(input("Enter third number: "))

if a <= b and a <= c:
    smallest = a
elif b <= a and b <= c:
    smallest = b
else:
    smallest = c

print("Smallest:", smallest)

Smallest: 55


### Problem 8.17: Number in Range (Easy)

Check if a number is between 1 and 100 (inclusive).

**Hint:** Check if 1 <= n <= 100.

In [22]:
# Your code here
n = int(input("Enter a number: "))

if 1 <= n <= 100:
    print("Number is in range")
else:
    print("Number is not in range")

Number is in range


### Problem 8.18: Absolute Value (Easy)

Calculate the absolute value of a number without using built-in functions.

**Hint:** If n < 0, return -n; else return n.

In [23]:
# Your code here
n = int(input("Enter a number: "))

if n < 0:
    n = -n

print("Absolute value:", n)

Absolute value: 2


### Problem 8.19: Days in Month (Medium)

Given month number (1-12), print the number of days. Consider leap year for February.

**Hint:** Use if-elif for each month. For Feb (month==2), check if leap year (use Problem 1.2 logic) to return 29 or 28.

In [24]:
# Your code here
month = int(input("Enter month number: "))
year = int(input("Enter year: "))

if month in [1, 3, 5, 7, 8, 10, 12]:
    print("31 days")

elif month in [4, 6, 9, 11]:
    print("30 days")

elif month == 2:

    if (year % 400 == 0) or (year % 4 == 0 and year % 100 != 0):
        print("29 days")
    else:
        print("28 days")

else:
    print("Invalid month")

31 days


### Problem 8.20: Profit or Loss Calculator (Easy)

Calculate profit or loss given cost price and selling price.

**Hint:** If selling_price > cost_price: profit = selling_price - cost_price. Else if less: loss. Else: no profit no loss.

In [25]:
# Your code here
cost_price = float(input("Enter cost price: "))
selling_price = float(input("Enter selling price: "))

if selling_price > cost_price:
    print("Profit:", selling_price - cost_price)

elif selling_price < cost_price:
    print("Loss:", cost_price - selling_price)

else:
    print("No profit, no loss")

Profit: 3110.0


### Problem 8.21: Uppercase or Lowercase (Easy)

Check if a character is uppercase or lowercase.

**Hint:** Use ch.isupper() or ch.islower(), or compare ASCII values.

In [26]:
# Your code here
ch = input("Enter a character: ")

if ch.isupper():
    print("Uppercase")
elif ch.islower():
    print("Lowercase")
else:
    print("Not a letter")

Lowercase


### Problem 8.22: Roots of Quadratic Equation (Hard)

Find the nature of roots of quadratic equation ax² + bx + c = 0:
- Real and distinct (discriminant > 0)
- Real and equal (discriminant = 0)
- Imaginary (discriminant < 0)

Discriminant = b² - 4ac

**Hint:** Calculate discriminant = b**2 - 4*a*c. Use if-elif-else to check if > 0, == 0, or < 0.

In [27]:
# Your code here
a = float(input("Enter a: "))
b = float(input("Enter b: "))
c = float(input("Enter c: "))

discriminant = b ** 2 - 4 * a * c

if discriminant > 0:
    print("Real and distinct roots")

elif discriminant == 0:
    print("Real and equal roots")

else:
    print("Imaginary roots")

Real and distinct roots


### Problem 8.23: Bonus Calculation (Medium)

Calculate employee bonus:
- Salary < 10000: 10% bonus
- Salary 10000-20000: 15% bonus
- Salary > 20000: 20% bonus
- If years of service > 5: Add extra 5%

**Hint:** Determine bonus percentage based on salary. If years > 5, add 5 to bonus percentage. Calculate: bonus = salary * bonus_percentage / 100.

In [28]:
# Your code here
salary = float(input("Enter salary: "))
years = int(input("Enter years of service: "))

if salary < 10000:
    bonus_percentage = 10
elif salary <= 20000:
    bonus_percentage = 15
else:
    bonus_percentage = 20

if years > 5:
    bonus_percentage += 5

bonus = salary * bonus_percentage / 100

print("Bonus:", bonus)

Bonus: 300.0


### Problem 8.24: Water State (Easy)

Determine water state based on temperature:
- Below 0°C: Ice
- 0-100°C: Water
- Above 100°C: Steam

**Hint:** If temp < 0: Ice, elif temp <= 100: Water, else: Steam.

In [29]:
# Your code here
temperature = float(input("Enter temperature: "))

if temperature < 0:
    print("Ice")
elif temperature <= 100:
    print("Water")
else:
    print("Steam")

Water


### Problem 8.25: Alphabet or Digit (Easy)

Check if a character is an alphabet or a digit.

**Hint:** Use ch.isalpha() to check alphabet, ch.isdigit() for digit.

In [30]:
# Your code here
ch = input("Enter a character: ")

if ch.isalpha():
    print("Alphabet")
elif ch.isdigit():
    print("Digit")
else:
    print("Special character")

Digit


### Problem 8.26: Tax Calculator (Medium)

Calculate income tax:
- Income 0-250000: No tax
- Income 250001-500000: 5% tax
- Income 500001-1000000: 10% tax
- Income above 1000000: 20% tax

**Hint:** Use if-elif-else to determine tax rate. Calculate: tax = income * tax_rate / 100.

In [31]:
# Your code here
income = float(input("Enter income: "))

if income <= 250000:
    tax_rate = 0
elif income <= 500000:
    tax_rate = 5
elif income <= 1000000:
    tax_rate = 10
else:
    tax_rate = 20

tax = income * tax_rate / 100

print("Tax:", tax)

Tax: 0.0


### Problem 8.27: Number of Days Calculator (Medium)

Calculate the difference in days between two dates (simplified: just day numbers in same month).

**Hint:** Simply calculate abs(day2 - day1) for absolute difference.

In [32]:
# Your code here
day1 = int(input("Enter first day: "))
day2 = int(input("Enter second day: "))

difference = abs(day2 - day1)

print("Difference in days:", difference)

Difference in days: 21


### Problem 8.28: Grade with Plus/Minus (Hard)

Assign grades with +/- modifiers:
- 97-100: A+
- 93-96: A
- 90-92: A-
- 87-89: B+
- (continue pattern for B, C, D, F)

**Hint:** Use if-elif-else with specific ranges. Start from highest (A+) and work down to F.

In [33]:
# Your code here
marks = int(input("Enter marks: "))

if 97 <= marks <= 100:
    grade = "A+"
elif 93 <= marks <= 96:
    grade = "A"
elif 90 <= marks <= 92:
    grade = "A-"
elif 87 <= marks <= 89:
    grade = "B+"
elif 83 <= marks <= 86:
    grade = "B"
elif 80 <= marks <= 82:
    grade = "B-"
elif 77 <= marks <= 79:
    grade = "C+"
elif 73 <= marks <= 76:
    grade = "C"
elif 70 <= marks <= 72:
    grade = "C-"
elif 67 <= marks <= 69:
    grade = "D+"
elif 63 <= marks <= 66:
    grade = "D"
elif 60 <= marks <= 62:
    grade = "D-"
elif 0 <= marks < 60:
    grade = "F"
else:
    grade = "Invalid marks"

print("Grade:", grade)

Grade: F


### Problem 8.29: Shipping Cost Calculator (Medium)

Calculate shipping cost based on weight and distance:
- Weight < 1kg: $5
- Weight 1-5kg: $10
- Weight > 5kg: $15
- If distance > 100km: Add $5

**Hint:** Determine base cost by weight. Then check if distance > 100 and add extra cost.

In [34]:
# Your code here
weight = float(input("Enter weight in kg: "))
distance = float(input("Enter distance in km: "))

if weight < 1:
    cost = 5
elif weight <= 5:
    cost = 10
else:
    cost = 15

if distance > 100:
    cost += 5

print("Shipping cost:", cost)

Shipping cost: 15


### Problem 8.30: Blood Donation Eligibility (Medium)

Check eligibility for blood donation:
- Age: 18-65
- Weight: >= 50kg
- Hemoglobin: >= 12.5 g/dL

**Hint:** Check if (18 <= age <= 65) and (weight >= 50) and (hemoglobin >= 12.5) all are True.

In [35]:
# Your code here
age = int(input("Enter age: "))
weight = float(input("Enter weight in kg: "))
hemoglobin = float(input("Enter hemoglobin level: "))

if 18 <= age <= 65 and weight >= 50 and hemoglobin >= 12.5:
    print("Eligible to donate blood")
else:
    print("Not eligible to donate blood")

Not eligible to donate blood


---

## Section 9: Additional For Loop Practice (30 Questions)

### Problem 9.1: Print Numbers 1 to N (Easy)

Print all numbers from 1 to n.

**Hint:** Use for i in range(1, n+1): print(i)

In [36]:
# Your code here
n = int(input("Enter n: "))

for i in range(1, n + 1):
    print(i)

1
2


### Problem 9.2: Print Numbers in Reverse (Easy)

Print numbers from n to 1 in descending order.

**Hint:** Use for i in range(n, 0, -1): print(i)

In [38]:
# Your code here
n = int(input("Enter n: "))

for i in range(n, 0, -1):
    print(i)

2
1


### Problem 9.3: Sum of First N Natural Numbers (Easy)

Calculate the sum: 1 + 2 + 3 + ... + n

**Hint:** Initialize sum=0. Loop from 1 to n+1, add each i to sum.

In [39]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):
    total += i

print("Sum:", total)

Sum: 3


### Problem 9.4: Product of Digits (Medium)

Calculate the product of all digits in a number.

**Hint:** Convert to string, loop through each digit: product *= int(digit). Or extract digits using modulo.

In [40]:
# Your code here
n = int(input("Enter a number: "))

product = 1

for digit in str(abs(n)):
    product *= int(digit)

print("Product:", product)

Product: 3


### Problem 9.5: Print All Odd Numbers (Easy)

Print all odd numbers from 1 to n.

**Hint:** Use range(1, n+1, 2) to get only odd numbers, or check if i%2 != 0.

In [41]:
# Your code here
n = int(input("Enter n: "))

for i in range(1, n + 1, 2):
    print(i)
    

1
3
5
7
9


### Problem 9.6: Print All Even Numbers (Easy)

Print all even numbers from 1 to n.

**Hint:** Use range(2, n+1, 2) to get only even numbers, or check if i%2 == 0.

In [43]:
# Your code here
n = int(input("Enter n: "))

for i in range(2, n + 1, 2):
    print(i)

2


### Problem 9.7: Count Vowels in String (Medium)

Count the number of vowels in a given string.

**Hint:** Loop through each character. Check if char.lower() in 'aeiou', increment count.

In [44]:
# Your code here
text = input("Enter a string: ")

count = 0

for ch in text:
    if ch.lower() in "aeiou":
        count += 1

print("Number of vowels:", count)

Number of vowels: 0


### Problem 9.8: Power Calculator (Easy)

Calculate x raised to power n using a for loop (don't use ** operator).

**Hint:** Initialize result=1. Loop n times: result *= x.

In [45]:
# Your code here
x = int(input("Enter base: "))
n = int(input("Enter power: "))

result = 1

for i in range(n):
    result *= x

print("Result:", result)

Result: 8


### Problem 9.9: Sum of Squares (Medium)

Calculate: 1² + 2² + 3² + ... + n²

**Hint:** Loop from 1 to n+1: sum += i**2 or i*i.

In [47]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):
    total += i * i

print("Sum of squares:", total)

Sum of squares: 42925


### Problem 9.10: Sum of Cubes (Medium)

Calculate: 1³ + 2³ + 3³ + ... + n³

**Hint:** Loop from 1 to n+1: sum += i**3 or i*i*i.

In [48]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):
    total += i * i * i

print("Sum of cubes:", total)

Sum of cubes: 216225


### Problem 9.11: Print Multiples of Number (Easy)

Print first 10 multiples of a given number.

**Hint:** Loop from 1 to 11: print(number * i).

In [50]:
# Your code here
number = int(input("Enter number: "))

for i in range(1, 11):
    print(number * i)

23
46
69
92
115
138
161
184
207
230


### Problem 9.12: Count Digits (Easy)

Count total number of digits in a number using for loop.

**Hint:** Convert to string: for digit in str(n): count += 1. Or use len(str(n)).

In [51]:
# Your code here
n = int(input("Enter number: "))

count = 0

for digit in str(abs(n)):
    count += 1

print("Number of digits:", count)

Number of digits: 1


### Problem 9.13: Find Factors (Medium)

Print all factors of a given number.

**Hint:** Loop from 1 to n+1. If n % i == 0, print i (it's a factor).

In [52]:
# Your code here
n = int(input("Enter number: "))

print("Factors:")

for i in range(1, n + 1):
    if n % i == 0:
        print(i)

Factors:
1
3
11
33


### Problem 9.14: Count Prime Numbers in Range (Hard)

Count how many prime numbers exist between 1 and n.

**Hint:** Use nested loops. Outer loop for each number, inner loop to check if prime. Count each prime found.

In [ ]:
# Your code here
n = int(input("Enter n: "))

count = 0

for num in range(2, n + 1):

    is_prime = True

    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            is_prime = False
            break
        
    if is_prime:
        count += 1

print("Number of primes:", count)

Number of primes: 8


### Problem 9.15: ASCII Values (Easy)

Print characters from 'A' to 'Z' with their ASCII values.

**Hint:** Loop from 65 to 91 (ASCII range). Use chr(i) to get character, i is ASCII value.

In [54]:
for i in range(65, 91):
    print(chr(i), "=", i)

A = 65
B = 66
C = 67
D = 68
E = 69
F = 70
G = 71
H = 72
I = 73
J = 74
K = 75
L = 76
M = 77
N = 78
O = 79
P = 80
Q = 81
R = 82
S = 83
T = 84
U = 85
V = 86
W = 87
X = 88
Y = 89
Z = 90


### Problem 9.16: Harmonic Series (Medium)

Calculate sum of harmonic series: 1 + 1/2 + 1/3 + ... + 1/n

**Hint:** Loop from 1 to n+1: sum += 1/i. Use floating point division.

In [55]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):
    total += 1 / i

print("Harmonic sum:", total)

Harmonic sum: 4.02724519543652


### Problem 9.17: Reverse String (Easy)

Reverse a string using for loop.

**Hint:** Loop backwards through string: for i in range(len(s)-1, -1, -1): reversed_str += s[i].

In [56]:
# Your code here
text = input("Enter a string: ")

reverse = ""

for i in range(len(text) - 1, -1, -1):
    reverse += text[i]

print("Reversed string:", reverse)

Reversed string: yaja


### Problem 9.18: Sum of Odd Numbers (Medium)

Calculate sum of all odd numbers from 1 to n.

**Hint:** Use range(1, n+1, 2) for odd numbers. Add each to sum.

In [57]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1, 2):
    total += i

print("Sum of odd numbers:", total)

Sum of odd numbers: 25


### Problem 9.19: Frequency Counter (Medium)

Count frequency of a specific digit in a number.

**Hint:** Convert number to string. Loop through each character, count if it matches target digit.

In [58]:
# Your code here
number = input("Enter a number: ")
digit = input("Enter digit to count: ")

count = 0

for ch in number:
    if ch == digit:
        count += 1

print("Frequency:", count)

Frequency: 1


### Problem 9.20: Alternating Sum (Medium)

Calculate: 1 - 2 + 3 - 4 + 5 - 6 + ... ± n

**Hint:** Loop from 1 to n+1. If i is odd: add i, if even: subtract i. Or use: sum += i if i%2==1 else -i.

In [59]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):

    if i % 2 == 1:
        total += i
    else:
        total -= i

print("Alternating sum:", total)

Alternating sum: -5


### Problem 9.21: Print Prime Factors (Hard)

Print all prime factors of a number.

**Hint:** Loop from 2 to n. For each i, while n % i == 0: print i, divide n by i. Continue with next i.

In [60]:
# Your code here
n = int(input("Enter number: "))

print("Prime factors:")

for i in range(2, n + 1):

    while n % i == 0:
        print(i, end=" ")
        n //= i

Prime factors:
2 5 

### Problem 9.22: Check Palindrome String (Medium)

Check if a string is a palindrome using for loop.

**Hint:** Reverse the string (Problem 9.17), then compare original with reversed.

In [61]:
# Your code here
text = input("Enter a string: ")

reverse = ""

for i in range(len(text) - 1, -1, -1):
    reverse += text[i]

if text == reverse:
    print("Palindrome")
else:
    print("Not a palindrome")

Not a palindrome


### Problem 9.23: Binary to Decimal (Hard)

Convert a binary number (as string) to decimal using for loop.

**Hint:** Loop through binary string. For each bit: decimal = decimal * 2 + int(bit).

In [62]:
# Your code here
binary = input("Enter binary number: ")

decimal = 0

for bit in binary:
    decimal = decimal * 2 + int(bit)

print("Decimal:", decimal)

Decimal: 2


### Problem 9.24: Decimal to Binary (Hard)

Convert a decimal number to binary using for loop.

**Hint:** While n > 0: append (n % 2) to result, then n = n // 2. Reverse the result.

In [64]:
# Your code here
n = int(input("Enter decimal number: "))

if n == 0:
    print("Binary: 0")
else:
    binary = ""

    while n > 0:
        binary = str(n % 2) + binary
        n //= 2

    print("Binary:", binary)

Binary: 10100


### Problem 9.25: Average of Numbers (Easy)

Calculate average of n numbers entered by user.

**Hint:** Loop n times to input numbers, add to sum. Average = sum / n.

In [66]:
# Your code here
n = int(input("How many numbers? "))

total = 0

for i in range(n):
    number = float(input("Enter number: "))
    total += number

average = total / n

print("Average:", average)

Average: 17.0


### Problem 9.26: Find Maximum (Easy)

Find the maximum among n numbers entered by user.

**Hint:** Initialize max with first number. Loop through remaining, update max if current number is larger.

In [68]:
# Your code here
n = int(input("How many numbers? "))

maximum = float(input("Enter number: "))

for i in range(n - 1):
    number = float(input("Enter number: "))

    if number > maximum:
        maximum = number

print("Maximum:", maximum)

Maximum: 222.0


### Problem 9.27: Find Minimum (Easy)

Find the minimum among n numbers entered by user.

**Hint:** Similar to Problem 9.26, but update min if current number is smaller.

In [69]:
# Your code here
n = int(input("How many numbers? "))

minimum = float(input("Enter number: "))

for i in range(n - 1):
    number = float(input("Enter number: "))

    if number < minimum:
        minimum = number

print("Minimum:", minimum)

Minimum: 11.0


### Problem 9.28: Compound Interest (Medium)

Calculate compound interest for n years: A = P(1 + r/100)^t

**Hint:** Use for loop to multiply principal by (1 + r/100) t times. Or use: A = P * (1 + r/100)**t.

In [70]:
# Your code here
principal = float(input("Enter principal: "))
rate = float(input("Enter rate: "))
years = int(input("Enter number of years: "))

amount = principal

for i in range(years):
    amount = amount * (1 + rate / 100)

compound_interest = amount - principal

print("Amount:", amount)
print("Compound Interest:", compound_interest)


Amount: 1488.3999999999999
Compound Interest: 488.39999999999986


### Problem 9.29: Perfect Numbers in Range (Hard)

Print all perfect numbers between 1 and n.

**Hint:** Outer loop for each number. For each, use inner loop to find sum of divisors (Problem 6.3). If sum equals number, print it.

In [74]:
n = int(input("Enter n: "))

print("Perfect numbers:")

for num in range(1, n + 1):

    total = 0

    for i in range(1, num):
        if num % i == 0:
            total += i

    if total == num:
        print(num)

Perfect numbers:
6
28


### Problem 9.30: Nested Factorial Sum (Hard)

Calculate: 1! + 2! + 3! + ... + n!

**Hint:** Outer loop from 1 to n+1. For each i, calculate factorial using inner loop (Problem 2.2), add to sum.

In [75]:
# Your code here
n = int(input("Enter n: "))

total = 0

for i in range(1, n + 1):

    factorial = 1

    for j in range(1, i + 1):
        factorial *= j

    total += factorial

print("Sum of factorials:", total)

Sum of factorials: 4037913


---

## Section 10: Additional While Loop Practice (30 Questions)

### Problem 10.1: Print 1 to N Using While (Easy)

Print numbers from 1 to n using while loop.

**Hint:** Initialize i=1. While i <= n: print(i), i += 1.

In [1]:
# Your code here
n  = int(input("Enter n: "))
i = 1

while i <= n:
    print(i)
    i += 1

1
2
3
4
5


### Problem 10.2: Countdown Timer (Easy)

Create a countdown from n to 1 using while loop.

**Hint:** While n > 0: print(n), n -= 1.

In [2]:
# Your code here
n  = int(input("Enter n: "))

while n > 0:
    print(n)
    n -= 1

3
2
1


### Problem 10.3: Find First Digit (Easy)

Find the first digit of a number using while loop.

**Hint:** While n >= 10: n = n // 10. When loop ends, n is the first digit.

In [3]:
# Your code here
number = int(input("Enter number: "))

while number >= 0:
    number = number // 10
print("First digit:", number)

KeyboardInterrupt: 

### Problem 10.4: Find Last Digit (Easy)

Find the last digit of a number.

**Hint:** Last digit = n % 10.

In [ ]:
# Your code here
n1  = int(input("Enter n: "))

last_digit = n1 % 10

print("Last digit:", last_digit)

### Problem 10.5: Remove Last Digit (Easy)

Remove the last digit from a number repeatedly until 0.

**Hint:** While n > 0: print(n), n = n // 10.

In [4]:
# Your code here
n = int(input("Enter number: "))

while n > 0:
    print(n)
    n = n // 10

2


### Problem 10.6: Sum Until Limit (Medium)

Keep adding numbers until sum exceeds a given limit.

**Hint:** i=1, sum=0. While sum <= limit: sum += i, i += 1. Print count or sum.

In [5]:
# Your code here
limit = int(input("Enter limit: "))

i = 1
total = 0
count = 0

while total <= limit:
    total += i
    i += 1
    count += 1

print("Sum:", total)
print("Count:", count)

Sum: 3
Count: 2


### Problem 10.7: Fibonacci Until N (Medium)

Print Fibonacci numbers until they exceed n.

**Hint:** a=0, b=1. While a <= n: print(a), calculate next (a, b = b, a+b).

In [7]:
# Your code here
n = int(input("Enter n: "))

a = 0
b = 1

while a <= n:
    print(a)
    a, b = b, a + b
print()

0
1
1
2
3



### Problem 10.8: Digit Separator (Medium)

Print each digit of a number on a separate line.

**Hint:** While n > 0: print(n % 10), n = n // 10. This prints in reverse; store in list to print in order.

In [9]:
n = int(input("Enter number: "))

digits = []

while n > 0:
    digit = n % 10
    digits.append(digit)
    n = n // 10

digits.reverse()

for digit in digits:
    print(digit)

3
3


### Problem 10.9: Factorial Using While (Medium)

Calculate factorial using while loop.

**Hint:** result=1, i=1. While i <= n: result *= i, i += 1.

In [8]:
# Your code here
n = int(input("Enter n: "))

result = 1
i = 1

while i <= n:
    result *= i
    i += 1

print("Factorial:", result)

Factorial: 3628800


### Problem 10.10: Power Using While (Medium)

Calculate x^n using while loop.

**Hint:** result=1, count=0. While count < n: result *= x, count += 1.

In [10]:
# Your code here
x = int(input("Enter base: "))
n = int(input("Enter power: "))

result = 1
count = 0

while count < n:
    result *= x
    count += 1

print("Result:", result)

Result: 27


### Problem 10.11: Menu-Driven Program (Medium)

Create a menu that keeps showing options until user selects exit.

**Hint:** while True: display menu, get choice. If exit chosen, break. Use if-elif for options.

In [11]:
# Your code here
while True:
    print("\n1. Add")
    print("2. Subtract")
    print("3. Exit")

    choice = int(input("Enter choice: "))

    if choice == 1:
        a = int(input("Enter first number: "))
        b = int(input("Enter second number: "))
        print("Result:", a + b)

    elif choice == 2:
        a = int(input("Enter first number: "))
        b = int(input("Enter second number: "))
        print("Result:", a - b)

    elif choice == 3:
        print("Program exited.")
        break

    else:
        print("Invalid choice")


1. Add
2. Subtract
3. Exit
Program exited.


### Problem 10.12: Validate Input (Medium)

Keep asking for input until user enters a valid positive number.

**Hint:** while True: get input. If n > 0, break. Otherwise continue loop.

In [12]:
# Your code here
while True:
    n = int(input("Enter a positive number: "))

    if n > 0:
        print("Valid number:", n)
        break
    else:
        print("Invalid input. Try again.")

Valid number: 2


### Problem 10.13: Password Attempts (Medium)

Allow 3 attempts to enter correct password.

**Hint:** attempts=0. While attempts < 3: get password. If correct, break. attempts += 1.

In [13]:
# Your code here
correct_password = "python123"

attempts = 0

while attempts < 3:
    password = input("Enter password: ")

    if password == correct_password:
        print("Login successful")
        break

    attempts += 1
    print("Incorrect password")

if attempts == 3:
    print("Account locked")

Incorrect password
Login successful


### Problem 10.14: GCD Using While (Medium)

Calculate GCD of two numbers using while loop.

**Hint:** Same as Problem 6.2. While b != 0: temp=b, b=a%b, a=temp.

In [15]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))

while b != 0:
    temp = b
    b = a % b
    a = temp

print("GCD:", a)


GCD: 22


### Problem 10.15: Remove Specific Digit (Hard)

Remove all occurrences of a specific digit from a number.

**Hint:** Extract each digit. If digit != target, build new number: result = result * 10 + digit.

In [16]:
n = int(input("Enter number: "))
target = int(input("Enter digit to remove: "))

result = 0
place = 1

while n > 0:
    digit = n % 10

    if digit != target:
        result = result + digit * place
        place *= 10

    n = n // 10

print("Result:", result)

Result: 6


### Problem 10.16: Keep Doubling (Easy)

Start with 1 and keep doubling until number exceeds n.

**Hint:** num=1. While num <= n: print(num), num *= 2.

In [17]:
# Your code here
n = int(input("Enter n: "))

num = 1

while num <= n:
    print(num)
    num *= 2

1
2


### Problem 10.17: Binary Conversion (Medium)

Convert decimal to binary using while loop.

**Hint:** Same as Problem 9.24. While n > 0: binary = str(n%2) + binary, n = n//2.

In [19]:
# Your code here
n = int(input("Enter decimal number: "))

binary = ""

if n == 0:
    binary = "0"

while n > 0:
    binary = str(n % 2) + binary
    n = n // 2

print("Binary:", binary)

Binary: 1011


### Problem 10.18: Octal Conversion (Medium)

Convert decimal to octal using while loop.

**Hint:** Similar to binary. While n > 0: octal = str(n%8) + octal, n = n//8.

In [18]:
# Your code here
n = int(input("Enter decimal number: "))

octal = ""

if n == 0:
    octal = "0"

while n > 0:
    octal = str(n % 8) + octal
    n = n // 8

print("Octal:", octal)

Octal: 70


### Problem 10.19: Hexadecimal Conversion (Hard)

Convert decimal to hexadecimal using while loop.

**Hint:** hex_digits = '0123456789ABCDEF'. While n > 0: hex_str = hex_digits[n%16] + hex_str, n = n//16.

In [20]:
# Your code here
n = int(input("Enter decimal number: "))

hex_digits = "0123456789ABCDEF"
hexadecimal = ""

if n == 0:
    hexadecimal = "0"

while n > 0:
    hexadecimal = hex_digits[n % 16] + hexadecimal
    n = n // 16

print("Hexadecimal:", hexadecimal)

Hexadecimal: 16


### Problem 10.20: Check Automorphic Number (Hard)

An automorphic number is a number whose square ends in the number itself.
Examples: 5 (25), 6 (36), 76 (5776)

**Hint:** Calculate square. Extract last digits of square (using %) equal to number of digits in n. Compare with n.

In [21]:
# Your code here
n = int(input("Enter number: "))

square = n * n
digits = len(str(n))

last_part = square % (10 ** digits)

if last_part == n:
    print("Automorphic number")
else:
    print("Not an automorphic number")


Automorphic number


### Problem 10.21: Neon Number (Medium)

A neon number is one where sum of digits of its square equals the number.
Example: 9 (9² = 81, 8+1 = 9)

**Hint:** Calculate square. Find sum of digits of square (Problem 3.2). Compare with original number.

In [22]:
# Your code here
n = int(input("Enter number: "))

square = n * n
sum_digits = 0

while square > 0:
    digit = square % 10
    sum_digits += digit
    square //= 10

if sum_digits == n:
    print("Neon number")
else:
    print("Not a neon number")

Not a neon number


### Problem 10.22: Spy Number (Medium)

A spy number is one where sum of digits equals product of digits.
Example: 123 (1+2+3 = 6, 1×2×3 = 6)

**Hint:** Calculate both sum and product of digits using while loop. Compare if equal.

In [23]:
# Your code here
n = int(input("Enter number: "))

temp = n
sum_digits = 0
product = 1

while temp > 0:
    digit = temp % 10

    sum_digits += digit
    product *= digit

    temp //= 10

if sum_digits == product:
    print("Spy number")
else:
    print("Not a spy number")

Not a spy number


### Problem 10.23: Harshad Number (Medium)

A Harshad number is divisible by the sum of its digits.
Example: 18 (1+8=9, 18÷9=2)

**Hint:** Calculate sum of digits. Check if original number % sum == 0.

In [24]:
# Your code he
n = int(input("Enter number: "))

temp = n
sum_digits = 0

while temp > 0:
    digit = temp % 10
    sum_digits += digit
    temp //= 10

if n % sum_digits == 0:
    print("Harshad number")
else:
    print("Not a Harshad number")

Not a Harshad number


### Problem 10.24: Abundant Number (Hard)

An abundant number is one where sum of proper divisors exceeds the number.
Example: 12 (1+2+3+4+6 = 16 > 12)

**Hint:** Find sum of divisors (1 to n-1). Check if sum > n.

In [25]:
# Your code here
n = int(input("Enter number: "))

i = 1
sum_divisors = 0

while i < n:
    if n % i == 0:
        sum_divisors += i

    i += 1

if sum_divisors > n:
    print("Abundant number")
else:
    print("Not an abundant number")

Not an abundant number


### Problem 10.25: Deficient Number (Hard)

A deficient number is one where sum of proper divisors is less than the number.
Example: 8 (1+2+4 = 7 < 8)

**Hint:** Find sum of divisors (1 to n-1). Check if sum < n.

In [27]:
# Your code here
n = int(input("Enter number: "))

i = 1
sum_divisors = 0

while i < n:
    if n % i == 0:
        sum_divisors += i

    i += 1

if sum_divisors < n:
    print("Deficient number")
else:
    print("Not a deficient number")

Deficient number


### Problem 10.26: Keep Dividing by 2 (Easy)

Keep dividing a number by 2 until it becomes odd.

**Hint:** While n % 2 == 0: n = n // 2. Print n when loop ends.

In [28]:
# Your code here
n = int(input("Enter number: "))

while n % 2 == 0:
    n = n // 2

print("Final odd value:", n)

Final odd value: 33


### Problem 10.27: Amicable Numbers (Hard)

Check if two numbers are amicable (sum of proper divisors of each equals the other).
Example: 220 and 284

**Hint:** Find sum of divisors of a (call it sum_a), sum of divisors of b (sum_b). Check if sum_a == b and sum_b == a.

In [29]:
# Your code here
a = int(input("Enter first number: "))
b = int(input("Enter second number: "))

sum_a = 0
i = 1

while i < a:
    if a % i == 0:
        sum_a += i
    i += 1

sum_b = 0
i = 1

while i < b:
    if b % i == 0:
        sum_b += i
    i += 1

if sum_a == b and sum_b == a:
    print("Amicable numbers")
else:
    print("Not amicable numbers")

Not amicable numbers


### Problem 10.28: Magic Number (Hard)

A magic number is one where repeatedly summing digits eventually gives 1.
Example: 19 → 1+9=10 → 1+0=1

**Hint:** While n > 9: calculate sum of digits, n = sum. After loop, check if n == 1.

In [31]:
# Your code here
n = int(input("Enter number: "))

while n > 9:
    sum_digits = 0

    while n > 0:
        digit = n % 10
        sum_digits += digit
        n //= 10

    n = sum_digits

if n == 1:
    print("Magic number")
else:
    print("Not a magic number")

Not a magic number


### Problem 10.29: ISBN Validator (Hard)

Validate a 10-digit ISBN number using the check digit algorithm.

**Hint:** ISBN-10: sum = (d1*10 + d2*9 + d3*8 + ... + d10*1). Valid if sum % 11 == 0. Loop with position weight.

In [32]:
# Your code here
isbn = input("Enter 10-digit ISBN: ")

if len(isbn) != 10 or not isbn.isdigit():
    print("Invalid ISBN")
else:
    total = 0
    i = 0

    while i < 10:
        digit = int(isbn[i])
        weight = 10 - i
        total += digit * weight
        i += 1

    if total % 11 == 0:
        print("Valid ISBN")
    else:
        print("Invalid ISBN")

Valid ISBN


### Problem 10.30: Kaprekar Number (Hard)

A Kaprekar number is one where square can be split into parts that sum to the original number.
Example: 45 (45² = 2025, 20+25 = 45)

**Hint:** Calculate square. Split into right part (last n digits) and left part (remaining). Check if left + right == original.

In [33]:
# Your code here
n = int(input("Enter number: "))

square = n * n
digits = len(str(n))

divisor = 10 ** digits

right = square % divisor
left = square // divisor

if left + right == n:
    print("Kaprekar number")
else:
    print("Not a Kaprekar number")

Not a Kaprekar number


---

## Section 11: Mixed Advanced Problems (10 Questions)

### Problem 11.1: Floyd's Triangle (Medium)

Print Floyd's triangle with n rows:
```
1
2 3
4 5 6
7 8 9 10
```

**Hint:** Use counter=1. Nested loops: for i (rows) and j (columns from 1 to i+1). Print counter, increment it.

In [34]:
# Your code here
n = int(input("Enter number of rows: "))

counter = 1

for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(counter, end=" ")
        counter += 1

    print()

1 
2 3 
4 5 6 
7 8 9 10 
11 12 13 14 15 
16 17 18 19 20 21 
22 23 24 25 26 27 28 
29 30 31 32 33 34 35 36 
37 38 39 40 41 42 43 44 45 
46 47 48 49 50 51 52 53 54 55 


### Problem 11.2: Hollow Square Pattern (Medium)

Print a hollow square pattern:
```
* * * * *
*       *
*       *
*       *
* * * * *
```

**Hint:** For first and last row: print all stars. For middle rows: print star, spaces, star. Check if i==0 or i==n-1 or j==0 or j==n-1.

In [35]:
# Your code here
n = int(input("Enter size: "))

for i in range(n):
    for j in range(n):

        if i == 0 or i == n - 1 or j == 0 or j == n - 1:
            print("*", end=" ")
        else:
            print(" ", end=" ")

    print()

* * 
* * 


### Problem 11.3: Spiral Matrix (Hard)

Print numbers in a spiral pattern for n×n matrix.

**Hint:** Use four boundaries (top, bottom, left, right). Fill in order: left to right (top row), top to bottom (right col), right to left (bottom row), bottom to top (left col). Shrink boundaries.

In [36]:
# Your code here
n = int(input("Enter matrix size: "))

matrix = []
num = 1

for i in range(n):
    row = []

    for j in range(n):
        row.append(0)

    matrix.append(row)

top = 0
bottom = n - 1
left = 0
right = n - 1

while top <= bottom and left <= right:

    # Left to right
    for j in range(left, right + 1):
        matrix[top][j] = num
        num += 1

    top += 1

    # Top to bottom
    for i in range(top, bottom + 1):
        matrix[i][right] = num
        num += 1

    right -= 1

    # Right to left
    if top <= bottom:
        for j in range(right, left - 1, -1):
            matrix[bottom][j] = num
            num += 1

        bottom -= 1

    # Bottom to top
    if left <= right:
        for i in range(bottom, top - 1, -1):
            matrix[i][left] = num
            num += 1

        left += 1


for row in matrix:
    print(row)

[1, 2]
[4, 3]


### Problem 11.4: Tic-Tac-Toe Game (Hard)

Create a simple two-player Tic-Tac-Toe game.

**Hint:** Use 3x3 list. Loop for turns. Display board, get position, check validity, place X or O. Check win conditions (rows, columns, diagonals) after each move.

In [37]:
# Your code here
board = [
    [" ", " ", " "],
    [" ", " ", " "],
    [" ", " ", " "]
]

def display_board():
    print()
    for row in board:
        print(" | ".join(row))
        print("-" * 9)


def check_winner(player):

    # Rows
    for row in board:
        if row[0] == player and row[1] == player and row[2] == player:
            return True

    # Columns
    for col in range(3):
        if (board[0][col] == player and
            board[1][col] == player and
            board[2][col] == player):
            return True

    # Diagonal
    if (board[0][0] == player and
        board[1][1] == player and
        board[2][2] == player):
        return True

    if (board[0][2] == player and
        board[1][1] == player and
        board[2][0] == player):
        return True

    return False


player = "X"

for turn in range(9):

    display_board()

    print("Player", player)

    row = int(input("Enter row (1-3): ")) - 1
    col = int(input("Enter column (1-3): ")) - 1

    if row < 0 or row > 2 or col < 0 or col > 2:
        print("Invalid position!")
        continue

    if board[row][col] != " ":
        print("Position already occupied!")
        continue

    board[row][col] = player

    if check_winner(player):
        display_board()
        print("Player", player, "wins!")
        break

    if player == "X":
        player = "O"
    else:
        player = "X"

else:
    display_board()
    print("Game Draw!")


  |   |  
---------
  |   |  
---------
  |   |  
---------
Player X

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O
Invalid position!

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O
Invalid position!

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O
Invalid position!

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O
Invalid position!

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O
Invalid position!

  | X |  
---------
  |   |  
---------
  |   |  
---------
Player O


ValueError: invalid literal for int() with base 10: ''

### Problem 11.5: Twin Primes (Medium)

Find all twin prime pairs between 1 and n. Twin primes differ by 2.
Example: (3,5), (5,7), (11,13)

**Hint:** Loop through numbers. For each, check if both i and i+2 are prime. If both prime, print pair (i, i+2).

In [38]:
# Your code here
n = int(input("Enter n: "))


def is_prime(num):

    if num < 2:
        return False

    i = 2

    while i * i <= num:
        if num % i == 0:
            return False

        i += 1

    return True


for i in range(2, n - 1):

    if is_prime(i) and is_prime(i + 2):
        print("(", i, ",", i + 2, ")")

### Problem 11.6: Matrix Addition (Medium)

Add two matrices (2D lists) entered by user.

**Hint:** Use nested loops. result[i][j] = matrix1[i][j] + matrix2[i][j] for each position.

In [39]:
# Your code here
rows = int(input("Enter number of rows: "))
cols = int(input("Enter number of columns: "))

matrix1 = []
matrix2 = []

print("Enter Matrix 1:")

for i in range(rows):
    row = []

    for j in range(cols):
        value = int(input(f"Enter value [{i}][{j}]: "))
        row.append(value)

    matrix1.append(row)


print("Enter Matrix 2:")

for i in range(rows):
    row = []

    for j in range(cols):
        value = int(input(f"Enter value [{i}][{j}]: "))
        row.append(value)

    matrix2.append(row)


result = []

for i in range(rows):
    row = []

    for j in range(cols):
        row.append(matrix1[i][j] + matrix2[i][j])

    result.append(row)


print("Result:")

for row in result:
    print(row)

Enter Matrix 1:
Enter Matrix 2:
Result:
[4, 4]
[34, 3]


### Problem 11.7: Matrix Transpose (Medium)

Find the transpose of a matrix.

**Hint:** Transpose means swap rows and columns. transpose[j][i] = matrix[i][j]. Use nested loops.

In [40]:
# Your code here

rows = int(input("Enter number of rows: "))
cols = int(input("Enter number of columns: "))

matrix = []

for i in range(rows):
    row = []

    for j in range(cols):
        value = int(input(f"Enter value [{i}][{j}]: "))
        row.append(value)

    matrix.append(row)


transpose = []

for j in range(cols):
    row = []

    for i in range(rows):
        row.append(matrix[i][j])

    transpose.append(row)


print("Transpose:")

for row in transpose:
    print(row)

Transpose:
[1, 1]
[2, 2]


### Problem 11.8: Sieve of Eratosthenes (Hard)

Implement the Sieve of Eratosthenes algorithm to find all primes up to n.

**Hint:** Create boolean list (all True). For each number i from 2 to √n, mark all multiples of i as False (not prime). Print numbers that remain True.

In [42]:
# Your code here
n = int(input("Enter n: "))

is_prime = [True] * (n + 1)

is_prime[0] = False
is_prime[1] = False

i = 2

while i * i <= n:

    if is_prime[i]:

        multiple = i * i

        while multiple <= n:
            is_prime[multiple] = False
            multiple += i

    i += 1


print("Prime numbers:")

for i in range(2, n + 1):

    if is_prime[i]:
        print(i, end=" ")

Prime numbers:
2 3 

### Problem 11.9: Anagram Checker (Medium)

Check if two strings are anagrams of each other.

**Hint:** Sort both strings and compare, or count frequency of each character in both strings and compare.

In [43]:
str1 = input("Enter first string: ")
str2 = input("Enter second string: ")

str1 = str1.lower().replace(" ", "")
str2 = str2.lower().replace(" ", "")

if sorted(str1) == sorted(str2):
    print("Anagram")
else:
    print("Not an anagram")

Not an anagram


### Problem 11.10: Sudoku Validator (Hard)

Validate if a given 9×9 Sudoku solution is correct.

**Hint:** Check three conditions: (1) Each row has unique 1-9, (2) Each column has unique 1-9, (3) Each 3x3 box has unique 1-9. Use sets to check uniqueness.

In [44]:
# Your code here
board = []

print("Enter 9 rows of Sudoku:")

for i in range(9):
    row = list(map(int, input().split()))
    board.append(row)


valid = True

# Check rows
for i in range(9):

    if set(board[i]) != set(range(1, 10)):
        valid = False
        break


# Check columns
if valid:

    for j in range(9):

        column = []

        for i in range(9):
            column.append(board[i][j])

        if set(column) != set(range(1, 10)):
            valid = False
            break


# Check 3x3 boxes
if valid:

    for row_start in range(0, 9, 3):

        for col_start in range(0, 9, 3):

            box = []

            for i in range(row_start, row_start + 3):

                for j in range(col_start, col_start + 3):
                    box.append(board[i][j])

            if set(box) != set(range(1, 10)):
                valid = False
                break

        if not valid:
            break


if valid:
    print("Valid Sudoku")
else:
    print("Invalid Sudoku")

Enter 9 rows of Sudoku:
Invalid Sudoku


---

## Submission Checklist

Before submitting, make sure:
- [ ] All problems are attempted
- [ ] Code is properly commented
- [ ] All code cells run without errors
- [ ] Outputs are displayed correctly
- [ ] Edge cases are handled
- [ ] Variable names are meaningful
- [ ] Code follows proper indentation

---

**Good luck with your assignments!**